picky-krb/src/crypto/diffie_hellman.rs

```diff
@@ -114,25 +119,43 @@ pub fn generate_key(
     nonce: Option<DhNonce>,
     cipher: &dyn Cipher,
 ) -> DiffieHellmanResult<Vec<u8>> {
-    let dh_shared_secret = generate_dh_shared_secret(public_key, private_key, modulus);
-
-    generate_key_from_shared_secret(&dh_shared_secret, nonce, cipher)
+    let dh_shared_secret = generate_dh_shared_secret(public_key, private_key, modulus)?;
+    generate_key_from_shared_secret(dh_shared_secret, nonce, cipher)
 }
 
 /// [Key and Parameter Requirements](https://www.rfc-editor.org/rfc/rfc2631#section-2.2)
 /// X9.42 requires that the private key x be in the interval [2, (q - 2)]
-pub fn generate_private_key<R: RngCore + CryptoRng>(q: &[u8], rng: &mut R) -> Vec<u8> {
-    let q = BigUint::from_bytes_be(q);
-
-    rng.gen_range(BigUint::from_bytes_be(&[2])..(q - 1_u32)).to_bytes_be()
+pub fn generate_private_key<R: TryCryptoRng>(q: &[u8], rng: &mut R) -> Result<Vec<u8>, R::Error> {
+    let q = BoxedUint::from_be_slice_vartime(q);
+    let low_bound = BoxedUint::from_be_slice_vartime(&[2]);
+    let high_bound = q - 1_u32;
+
+    let min_bits = low_bound.bits();
+    let max_bits = high_bound.bits();
+    loop {
```

Comment:
`crypto_bigint::BoxedUint` does not implement that trait, which can generate a random number in a range.
So, I had to rewrite the old logic by doing a little bit of brute force.